src/WebJobs.Extensions.DurableTask/ContextImplementations/DurableClient.cs

```diff
@@ -491,12 +491,45 @@ async Task<PurgeHistoryResult> IDurableOrchestrationClient.PurgeInstanceHistoryA
             return this.DurabilityProvider.GetOrchestrationStateWithPagination(condition, cancellationToken);
         }
 
+        private static EntityQueryResult ConvertToEntityQueryResult(IEnumerable<DurableEntityStatus> entities, string continuationToken)
+        {
+            var entityQueryResult = new EntityQueryResult();
+            entityQueryResult.Entities = entities;
+            entityQueryResult.ContinuationToken = continuationToken;
+
+            return entityQueryResult;
+        }
+
         /// <inheritdoc />
         async Task<EntityQueryResult> IDurableEntityClient.ListEntitiesAsync(EntityQuery query, CancellationToken cancellationToken)
         {
+            if (!query.IncludeDeleted)
+            {
+                query.FetchState = true;
+            }
+
             var condition = new OrchestrationStatusQueryCondition(query);
             var result = await ((IDurableClient)this).ListInstancesAsync(condition, cancellationToken);
             var entityResult = new EntityQueryResult(result);
+
+            if (!query.IncludeDeleted)
+            {
+                var statefulEntities = entityResult.Entities.Where(e => e.State != null);
+                var statefulEntityResult = ConvertToEntityQueryResult(statefulEntities, entityResult.ContinuationToken);
+
+                while (statefulEntities.ToList().Count < query.PageSize && !string.IsNullOrEmpty(entityResult.ContinuationToken))
```

Comment:
There's no benefit to doing a `.ToList()` here since you're not actually doing anything with the list you're creating. It would have been better to just say `statefulEntities.Count()`.